src/Streamly/Prelude.hs

```diff
@@ -361,6 +379,20 @@ notElem e m = go (toStream m)
 length :: (Streaming t, Monad m) => t m a -> m Int
 length = foldl (\n _ -> n + 1) 0 id
 
+-- | Returns the elements of the stream in reverse order. 
+-- The stream must be finite.
+reverse :: (Streaming t) => t m a -> t m a
+reverse m = fromStream $ go Nothing (toStream m)
+    where
+      go rev rest = Stream $ \s stp yld ->
+          let
+              stop             = case rev of
+                  Nothing -> runStream snil s stp yld
+                  Just str -> runStream str s stp yld
+              yield a Nothing  = runStream (a `scons` rev) s stp yld
+              yield a (Just x) = runStream (go (Just $ a `scons` rev) x) s stp yld
+           in runStream rest Nothing stop yield
```

Comment:
We need to pass on the `SVar` `s` here instead of using `Nothing`. I need to figure out a way to add quick tests for such concurrency related stuff.